src/transactions/SponsorshipUtils.cpp

```diff
@@ -0,0 +1,780 @@
+    if (acc.numSubEntries >= ACCOUNT_SUBENTRY_LIMIT)
+    {
+        return SponsorshipResult::TOO_MANY_SUBENTRIES;
+    }
+
+    return canEstablishSignerSponsorship(lh, signerIt, sponsoringAcc,
+                                         sponsoredAcc);
+}
+
+// TODO(jonjove): Use signerIt or remove it
+void
+canRemoveSignerWithoutSponsorship(
+    LedgerHeader const& lh, std::vector<Signer>::const_iterator const& signerIt,
+    LedgerEntry const& acc)
+{
+    if (acc.data.account().numSubEntries < 1)
+    {
+        throw std::runtime_error("invalid account state");
+    }
+}
+
+// TODO(jonjove): Use signerIt or remove it
+void
+canRemoveSignerWithSponsorship(
+    LedgerHeader const& lh, std::vector<Signer>::const_iterator const& signerIt,
+    LedgerEntry const& sponsoringAcc, LedgerEntry const& sponsoredAcc)
+{
+    if (lh.ledgerVersion < 14)
+    {
+        throw std::runtime_error("sponsorship before version 14");
+    }
+
+    if (getNumSponsoring(sponsoringAcc) < 1)
+    {
+        throw std::runtime_error("invalid sponsoring account state");
+    }
+
+    if (sponsoredAcc.data.account().numSubEntries < 1 ||
+        getNumSponsored(sponsoredAcc) < 1)
+    {
+        throw std::runtime_error("invalid sponsored account state");
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// Utility functions to create/remove with/without sponsorship
+//
+////////////////////////////////////////////////////////////////////////////////
+void
+createEntryWithoutSponsorship(LedgerEntry& le, LedgerEntry& acc)
+{
+    // TODO(jonjove): Handle ClaimableBalance
+    if (le.data.type() != ACCOUNT)
+    {
+        ++acc.data.account().numSubEntries;
+    }
+}
+
+void
+createEntryWithSponsorship(LedgerEntry& le, LedgerEntry& sponsoringAcc,
+                           LedgerEntry* sponsoredAcc)
+{
+    if (sponsoredAcc)
+    {
+        createEntryWithoutSponsorship(le, *sponsoredAcc);
+    }
+    establishEntrySponsorship(le, sponsoringAcc, sponsoredAcc);
+}
+
+void
+removeEntryWithoutSponsorship(LedgerEntry& le, LedgerEntry& acc)
+{
+    if (le.data.type() != ACCOUNT)
+    {
+        --acc.data.account().numSubEntries;
+    }
+}
+
+void
+removeEntryWithSponsorship(LedgerEntry& le, LedgerEntry& sponsoringAcc,
+                           LedgerEntry* sponsoredAcc)
+{
+    if (sponsoredAcc)
+    {
+        removeEntryWithoutSponsorship(le, *sponsoredAcc);
+    }
+    removeEntrySponsorship(le, sponsoringAcc, sponsoredAcc);
+}
+
+void
+createSignerWithoutSponsorship(
+    std::vector<Signer>::const_iterator const& signerIt, LedgerEntry& acc)
+{
+    auto& ae = acc.data.account();
+    ++ae.numSubEntries;
+    if (ae.ext.v() == 1 && ae.ext.v1().ext.v() == 2)
+    {
+        size_t n = signerIt - ae.signers.begin();
+        auto& extV2 = ae.ext.v1().ext.v2();
+        extV2.signerSponsoringIDs.insert(extV2.signerSponsoringIDs.begin() + n,
+                                         {});
```

Comment:
Good idea, this looks strange. I have added a comment here. It is also on my to-do list to add some general comments in this file about what to expect from various functions. But I want to emphasize that we are actually adding a null `signerSponsoringID` rather than a null `signer`.